### ISSUE TITLE
[Bug]: Competitions editing a competition all users disappeared

### ISSUE THREAD

**Issue by jonsnowpt** [Betarena/scores#1579]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### 🐞 Describe the bug

When editing a competition to finish, all users disappeared.

![Screenshot 2023-09-05 at 06 54 44](https://github.com/Betarena/scores/assets/37311649/51c18f95-1fc4-4c3f-a915-55a5932d251c)
![Screenshot 2023-09-05 at 06 54 40](https://github.com/Betarena/scores/assets/37311649/33753439-c161-4466-8251-a0cdd84f1136)

### 🤔 Expected Behavior

Editing a competition should never delete the participants.

### What environment is the bug/issue detected in.

- [ ] Production (scores.betarena.com)
- [X] Development

### What browsers are you seeing the problem on?

- [X] Firefox
- [ ] Chrome
- [ ] Safari
- [ ] Microsoft Edge
- [ ] Brave
- [ ] Opera

### What device/platform can the problem be seen?

- [ ] 📱 Mobile
- [ ] 💻 Tablet
- [X] 🖥️ Laptop/Desktop

### Steps To Reproduce

The competition fixture status was not fit and after changing it, the participants disappeared.

**Comment by migbash** [comment]
@jonsnowpt 

This is strange, and has never occurred yet, the backend does not update the participants when the `data` column updates. It only recalculates necessary fields based on formulas.

I believe the issue has happened because you most likely had the target competition tab opened on your browser before hand for target competition, for a time when there where no participants in the competition, and upon `save` it has overwritten the data, erasing participants.

**Comment by jonsnowpt** [comment]
@migbash yes, it might be. We can close the issue for now and I will keep checking. Thanks

**Comment by migbash** [comment]
> **Warning**
> This is a general issue that needs to be addressed, to prevent participants from `disappearing` by **accident**. Currently, we have a partial solution already implemented for this exact situation via the `participant_version` column data on `competitions/data` table, which makes sure when new `participants` are added, the `participant_versions` are the same and incremented (a.k.a optimistic locking). However, when updating the `data` column, this `participant_version` check cannot be added as a `constraint` to have it be `participant_version >= participant_version`. Instead, `ON UPDATE ...` SQL can be added to check for this `NEW` vs `OLD` values, and act accordingly. This way, the direct tampering of the `data` column will also be restricted to dangerous data mutations, and this will solve such accidental cases, that will ensure correct data state is being updated.

- [🔗 | SQL Constraint to prevent updating a column based on prior value](https://stackoverflow.com/questions/14367332/sql-constraint-to-prevent-updating-a-column-based-on-its-prior-value)
- [🔗 | plpgsql trigger documentation](https://www.postgresql.org/docs/current/plpgsql-trigger.html)
